### ISSUE TITLE
Does the neutrinordp proxy module support connection to ipv6 windows resources?

### ISSUE THREAD

**Issue by Jackobama** [neutrinolabs/xrdp#1745]
Hello, I would like to ask whether the neutrinordp proxy module supports connecting to ipv6 windows resources

**Comment by matt335672** [comment]
It's a good question.

The neutrinordp proxy depends on the fork of freerdp at https://github.com/neutrinolabs/NeutrinoRDP

This has very limited IPv6 support. The code in `tcp_connect()` uses the right functions to connect to IPv6 hosts, but the function `tcp_get_ip_address()` to get the IP address from the connection looks to me like it only supports IPv4.

In other words I can't fully answer your question by looking at code. Have you tried this and found it doesn't work, or are you contemplating trying it?

**Comment by Jackobama** [comment]
> It's a good question.
> 
> The neutrinordp proxy depends on the fork of freerdp at https://github.com/neutrinolabs/NeutrinoRDP
> 
> This has very limited IPv6 support. The code in `tcp_connect()` uses the right functions to connect to IPv6 hosts, but the function `tcp_get_ip_address()` to get the IP address from the connection looks to me like it only supports IPv4.
> 
> In other words I can't fully answer your question by looking at code. Have you tried this and found it doesn't work, or are you contemplating trying it?


Thank you very much for your answer, I have tried to use the neutrinordp proxy module to connect to the windows resources of ipv6, but the connection failed.
The log of my connection failure is as follows：
lfreerdp_pre_connect:
The colormap is all NULL
host fe80::70f7:24be:91ee:7ea9 port 3389
unable to connect to fe80::70f7:24be:91ee:7ea9:3389
Error: protocol security negotiation failure
lxrdp_connect: freerdp_connect returned 0
Failure to connect
[20201209-10:20:47] [INFO ] freerdp_connect Failed to destination :fe80::70f7:24be:91ee:7ea9:3389
[20201209-10:20:47] [ERROR] xrdp_wm_log_msg: Error connecting to:

**Comment by matt335672** [comment]
I've had a play with this, and I might be able to get you a little further.

fe80::70f7:24be:91ee:7ea9 is a link-local address, and so you need to tell the system which interface it corresponds to.

On my test system, I've got the following network setup (I've snipped my GUAs):-

```
$ ip -6 addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 fe80::f607:6a35:5cc8:68dd/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
```

I knocked up a quick test program with the guts of `tcp_connect()` in it. I can post it if you like, but in essence it takes a host and tries to connect to port 22 (the ssh port).

Results of running this are as follows:-

<pre>
$ ./temp fe80::f607:6a35:5cc8:68dd
host fe80::f607:6a35:5cc8:68dd port 22
unable to connect to fe80::f607:6a35:5cc8:68dd:22
0
$ ./temp fe80::f607:6a35:5cc8:68dd<b>%2</b>
host fe80::f607:6a35:5cc8:68dd%2 port 22
connected to fe80::f607:6a35:5cc8:68dd%2:22
1
./temp fe80::f607:6a35:5cc8:68dd<b>%ens3</b>
host fe80::f607:6a35:5cc8:68dd%ens3 port 22
connected to fe80::f607:6a35:5cc8:68dd%ens3:22
1
</pre>

In other words, appending the interface name or zone index to your IP address is necessary with IPv6 link-local addresses - here's a [wikipedia](https://en.wikipedia.org/wiki/IPv6_address#zone_index) reference which I think explains it better than I have.

Alternatively, use an IPv6 address with global scope rather than link scope.

I hope that's useful. Let me know if I've not been clear.

**Comment by Jackobama** [comment]
> I've had a play with this, and I might be able to get you a little further.
> 
> fe80::70f7:24be:91ee:7ea9 is a link-local address, and so you need to tell the system which interface it corresponds to.
> 
> On my test system, I've got the following network setup (I've snipped my GUAs):-
> 
> ```
> $ ip -6 addr
> 1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000
>     inet6 ::1/128 scope host 
>        valid_lft forever preferred_lft forever
> 2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
>     inet6 fe80::f607:6a35:5cc8:68dd/64 scope link noprefixroute 
>        valid_lft forever preferred_lft forever
> ```
> 
> I knocked up a quick test program with the guts of `tcp_connect()` in it. I can post it if you like, but in essence it takes a host and tries to connect to port 22 (the ssh port).
> 
> Results of running this are as follows:-
> 
> $ ./temp fe80::f607:6a35:5cc8:68dd
> host fe80::f607:6a35:5cc8:68dd port 22
> unable to connect to fe80::f607:6a35:5cc8:68dd:22
> 0
> $ ./temp fe80::f607:6a35:5cc8:68dd%2
> host fe80::f607:6a35:5cc8:68dd%2 port 22
> connected to fe80::f607:6a35:5cc8:68dd%2:22
> 1
> ./temp fe80::f607:6a35:5cc8:68dd%ens3
> host fe80::f607:6a35:5cc8:68dd%ens3 port 22
> connected to fe80::f607:6a35:5cc8:68dd%ens3:22
> 1
> In other words, appending the interface name or zone index to your IP address is necessary with IPv6 link-local addresses - here's a [wikipedia](https://en.wikipedia.org/wiki/IPv6_address#zone_index) reference which I think explains it better than I have.
> 
> Alternatively, use an IPv6 address with global scope rather than link scope.
> 
> I hope that's useful. Let me know if I've not been clear.

Thank you again for your answer. Your reply is very important to me. I have used the neutrinordp proxy module to connect to the ipv6 windows resources. The reason for the failure before is that my ipv6 address did not add the network card name. Now I add it. , I can connect, I really appreciate your spare time to answer my questions.

**Comment by matt335672** [comment]
That's good news.

Thanks very much for reporting back. It's very useful to other users who may have a similar issue.

I'll close this issue for now then.

**Comment by Jackobama** [comment]
> That's good news.
> 
> Thanks very much for reporting back. It's very useful to other users who may have a similar issue.
> 
> I'll close this issue for now then.

Ok, thank you for your help.
